src/uct/ib/base/ib_iface.c

```diff
@@ -382,37 +382,35 @@ ucs_status_t uct_ib_iface_create_ah(uct_ib_iface_t *iface,
                                           uct_ib_iface_md(iface)->pd, ah_p);
 }
 
-static ucs_status_t uct_ib_iface_init_pkey(uct_ib_iface_t *iface,
-                                           const uct_ib_iface_config_t *config)
+static ucs_status_t uct_ib_iface_find_pkey(uct_ib_iface_t *iface,
+                                           unsigned user_pkey_value)
 {
-    uct_ib_device_t *dev = uct_ib_iface_device(iface);
+    uct_ib_device_t *dev  = uct_ib_iface_device(iface);
     uint16_t pkey_tbl_len = uct_ib_iface_port_attr(iface)->pkey_tbl_len;
     uint16_t pkey_index, port_pkey, pkey;
 
-    if (config->pkey_value > UCT_IB_PKEY_PARTITION_MASK) {
-        ucs_error("Requested pkey 0x%x is invalid, should be in the range 0..0x%x",
-                  config->pkey_value, UCT_IB_PKEY_PARTITION_MASK);
-        return UCS_ERR_INVALID_PARAM;
-    }
-
-    /* get the user's pkey value and find its index in the port's pkey table */
     for (pkey_index = 0; pkey_index < pkey_tbl_len; ++pkey_index) {
         /* get the pkey values from the port's pkeys table */
-        if (ibv_query_pkey(dev->ibv_context, iface->config.port_num, pkey_index,
-                           &port_pkey))
+        if (ibv_query_pkey(dev->ibv_context, iface->config.port_num,
+                           pkey_index, &port_pkey))
         {
             ucs_error("ibv_query_pkey("UCT_IB_IFACE_FMT", index=%d) failed: %m",
                       UCT_IB_IFACE_ARG(iface), pkey_index);
         }
 
         pkey = ntohs(port_pkey);
         if (!(pkey & UCT_IB_PKEY_MEMBERSHIP_MASK)) {
-            ucs_debug("skipping send-only pkey[%d]=0x%x", pkey_index, pkey);
+            /* if pkey = 0x0, just skip it w/o debug trace, because 0x0
+             * means that there is no real pkey configured at this index */
+            if (pkey) {
+                ucs_debug("skipping send-only pkey[%d]=0x%x", pkey_index, pkey);
+            }
             continue;
         }
 
-        /* take only the lower 15 bits for the comparison */
-        if ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey_value) {
+        if ((user_pkey_value == UCT_IB_PKEY_ANY) ||
```

Comment:
this change seems unnecessarily complicated; can't we just modify the previous code something like this:
```
  if ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey_value || !pkey_already_set) {
    iface->pkey_index = pkey_index;
    ...
  }
```
instead of making a new function, calling it with UCT_IB_PKEY_ANY on 2nd time etc..

and modify the error message from "the requested pkey cannot be used" to "no valid pkey is found on device:port..."

Reply:
@yosefe anyway we need to go through the pkey table twice (1- check table, 2 - take the first pkey with full membership):
```
    pkey_set_first = 0;
loop:
    for (pkey_index = 0; pkey_index < pkey_tbl_len; ++pkey_index) {
        ...
        if (((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey_value) ||
            pkey_set_first) {
            ...
        }
    }
    /* If not found - set the first available pkey */
    pkey_set_first = 1;
    goto loop;
```

it is almost the same as your idea, but adds `loop:` label and `goto loop` operator to take the first pkey
wdyt?

Reply:
implemented @yosefe's approach